src/test/java/picard/sam/RevertSamTest.java

```diff
@@ -592,4 +606,189 @@ public void testHardClipRoundTrip() throws Exception {
+        try (PrintWriter writer = new PrintWriter(Files.newBufferedWriter(outputTablePath))) {
+            // Write header
+            writeRow(writer, RevertSam.READ_GROUP_ID_COLUMN_NAME, RevertSam.OUTPUT_COLUMN_NAME);
+
+            for (int i = 0; i < readNames.size(); i++){
+                writeRow(writer, readNames.get(i), outputFiles.get(i));
+            }
+        } catch (IOException e) {
+            throw new PicardException("Error writing a tsv file to " + outputTablePath, e);
+        }
+    }
+
+    // Write to a file (writer) the given array of strings as a tab-delimited row
+    private static void writeRow(final PrintWriter writer, final String... row) {
+        final int delimiter = '\t';
+        for (int j = 0; j < row.length; j++){
+            writer.write(row[j]);
+            if (j < row.length - 1) {
+                writer.write(delimiter);
+            } else {
+                writer.println();
+            }
+        }
+    }
+
+    private List<SAMReadGroupRecord> getReadGroups(final Path sam){
+        final List<SAMReadGroupRecord> readGroupsInInput = SamReaderFactory.makeDefault().open(sam).getFileHeader().getReadGroups();
+        return readGroupsInInput;
+    }
+
+    /***
+     *
+     * @param inputSAM The path to the input SAM file
+     * @param outputPath The path to a directory, when output by read group is true and the map isn't provided.
+     *                   The path to a single output SAM file, when output by read group is turned off. May be null.
+     * @param outputByReadGroup Whether to output by read group
+     * @param outputMapOptions Dictates whether to dynamically create an output map and where (local or cloud)
+     */
+    @Test(dataProvider = "cloudTestData", groups = "cloud")
+    public void testCloud(final PicardHtsPath inputSAM, final PicardHtsPath outputPath, final boolean outputByReadGroup,
+                          final OutputMapOptions outputMapOptions, final PicardHtsPath reference) {
+        final List<SAMReadGroupRecord> readGroupsInInput = SamReaderFactory.makeDefault().open(inputSAM.toPath()).getFileHeader().getReadGroups();
+        final List<String> args = new ArrayList<>(Arrays.asList(
+                "INPUT=" + inputSAM.getURIString(),
+                "OUTPUT_BY_READGROUP=" + outputByReadGroup));
+
+        // Add a prefix to avoid name collisions across PRs etc.
+        String prefix = null;
+        if (outputPath != null){
+            args.add("OUTPUT=" + outputPath.getURIString());
+            if (outputByReadGroup){
+                LocalDateTime now = LocalDateTime.now();
+                DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd_HH:mm:ss.SSS");
+                prefix = now.format(formatter);
+                args.add("PREFIX=" + prefix);
```

Comment:
Unfortunately, I think this name collision resolution strategy is a step backwards.

Instead of using DEFAULT_CLOUD_TEST_OUTPUT_DIR for the output directory in these tests, it would be much simpler to change the test code to create a new temporary directory/bucket in the staging dir, and pass that to RevertSam as the outputPath. Then you don't need to add the new tool arg PREFIX, which is really only useful for tests, or this prefix/timestamp code (which is still theoretically subject to name collision).

If there are test cases for which this (or copying the inputs - see the discussion in the engine team space) doesn't suffice for some reason, let me know and we can think about other alternatives. But I'd very much prefer to not add args to tools that are only useful for tests if we can possibly avoid it.

Reply:
Yes that would be the most natural thing to do, except that there are no directories in google cloud so we can't "create" a temp directory…what's confusing is that in the browser, they do let you create a directory (that's how I created DEFAULT_CLOUD_TEST_OUTPUT_DIR). For instance `createDirectory` method in `CloudStorageFileSystemProvider.java` is a no-op. 

Come to think of it, as long as we don't check that the output directory exists (which we currently do not when the directory is in the cloud), I could probably just create an nonexistent random directory name as OUTPUT, and the files will be created under it. I will do some digging and let you know.

Reply:
@tsato - you can create a cloud directory (see the engine-team slack discussion from Tuesday 7/9 about this). Also, I'm currently doing a similar thing myself in a different PR so I can share that code with you if you like.

Reply:
Let me know if my approach in the new commit is what you had in mind.